File: verify_i18n.mjs
import { readFileSync, readdirSync, statSync } from 'fs';
import { join } from 'path';

const NOISE = new Set(['.', 'script', 'meta', 'link', '：', '']);
const keys = new Set();
function walk(dir) {
  for (const f of readdirSync(dir)) {
    const p = join(dir, f);
    if (statSync(p).isDirectory()) walk(p);
    else if (f.endsWith('.jsx') || f.endsWith('.js')) {
      const txt = readFileSync(p, 'utf8');
      for (const m of txt.matchAll(/t\(['"]([^'"]+)['"]\)/g)) {
        const k = m[1];
        if (NOISE.has(k)) continue;
        keys.add(k);
      }
    }
  }
}
walk('src');

const mod = await import('./src/i18n/translations.js');
const trans = mod.translations;

function getPath(obj, path) {
  return path.split('.').reduce((o, k) => (o == null ? undefined : o[k]), obj);
}

const missZh = [...keys].filter(k => getPath(trans.zh, k) === undefined);
const missEn = [...keys].filter(k => getPath(trans.en, k) === undefined);

// 真正的 photo card key：在程式裡用 tKey + 'Desc' 模式
// 從 jsx 抓 `tKey: 'xxx'` 再檢查 xxx 與 xxxDesc 都存在
const photoSet = new Set();
for (const f of readdirSync('src/pages')) {
  const txt = readFileSync(join('src/pages', f), 'utf8');
  for (const m of txt.matchAll(/tKey:\s*['"]([^'"]+)['"]/g)) {
    photoSet.add(m[1]);
  }
}
const descZh = [...photoSet].filter(k => getPath(trans.zh, k) === undefined || getPath(trans.zh, k+'Desc') === undefined);
const descEn = [...photoSet].filter(k => getPath(trans.en, k) === undefined || getPath(trans.en, k+'Desc') === undefined);

console.log('總 t() key 數:', keys.size);
console.log('photo card tKey 數:', photoSet.size);
console.log('zh 缺失:', missZh);
console.log('en 缺失:', missEn);
console.log('zh photo Desc 缺失:', descZh);
console.log('en photo Desc 缺失:', descEn);
console.log((missZh.length||missEn.length||descZh.length||descEn.length) ? 'HAS MISSING' : 'OK ✅ 所有 key 雙語齊全');

File: verify_i18n.py
import re, os

src_dir = 'src'
keys = set()
for root, _, files in os.walk(src_dir):
    for fn in files:
        if fn.endswith(('.jsx', '.js')):
            with open(os.path.join(root, fn), encoding='utf-8') as fh:
                txt = fh.read()
            for m in re.findall(r"t\(['\"]([^'\"]+)['\"]\)", txt):
                keys.add(m)

# 讀 translations.js，去掉 // 註解與 export，改為可 exec
raw = open(os.path.join(src_dir, 'i18n', 'translations.js'), encoding='utf-8').read()
lines = []
for line in raw.splitlines():
    if line.strip().startswith('//'):
        continue
    lines.append(line.replace('export const', 'const'))
src = '\n'.join(lines)
ns = {}
exec(src + '\nns["translations"]=translations', ns)
trans = ns['translations']

missing_zh = [k for k in keys if k not in trans['zh']]
missing_en = [k for k in keys if k not in trans['en']]
# photo desc 子鍵
desc_missing_zh = [k+'Desc' for k in keys if k in trans['zh'] and (k+'Desc') not in trans['zh']]
desc_missing_en = [k+'Desc' for k in keys if k in trans['en'] and (k+'Desc') not in trans['en']]

print('總 t() key 數:', len(keys))
print('zh 缺失:', missing_zh)
print('en 缺失:', missing_en)
print('zh photo Desc 缺失:', desc_missing_zh)
print('en photo Desc 缺失:', desc_missing_en)
print('OK' if not (missing_zh or missing_en or desc_missing_zh or desc_missing_en) else 'HAS MISSING')
